src/esphome/climatedevice/climatedevice.h

```diff
@@ -0,0 +1,232 @@
+#ifndef ESPHOME_CLIMATEDEVICE_CLIMATEDEVICE_H
+#define ESPHOME_CLIMATEDEVICE_CLIMATEDEVICE_H
+
+#include "esphome/defines.h"
+
+#ifdef USE_CLIMATEDEVICE
+
+#include "esphome/component.h"
+#include "esphome/automation.h"
+#include "esphome/esppreferences.h"
+#include "esphome/climatedevice/climatedevice_traits.h"
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace climatedevice {
+
+/// Simple enum to represent the mode of a climate device.
+enum ClimateDeviceMode {
+  CLIMATEDEVICE_MODE_OFF = 0,   ///< The climate device is off.
+  CLIMATEDEVICE_MODE_AUTO = 1,  ///< The climate device is in auto mode.
+  CLIMATEDEVICE_MODE_COOL = 2,  ///< The climate device is in cool mode.
+  CLIMATEDEVICE_MODE_HEAT = 3,  ///< The climate device is in heat mode.
+};
+
+template<typename... Ts> class ControlAction;
+template<typename... Ts> class PublishAction;
+template<typename... Ts> class CurrentTemperatureAction;
+
+struct ClimateDeviceState {
+  /// The mode of the climate device.
+  ClimateDeviceMode mode{CLIMATEDEVICE_MODE_OFF};
+  /// The target temperature.
+  float target_temperature{NAN};
+};
+
+#ifdef USE_MQTT_CLIMATEDEVICE
+class MQTTClimateDeviceComponent;
+#endif
+
+/** This class is climate device base class between the hardware backend and the MQTT frontend.
+ *
+ * A climate device has several variables that determine the current state: mode
+ * (OFF, AUTO, HEAT, COOL), current_temperature, target_temperature and traits (what features are supported).
+ * Both the frontend and the backend can register callbacks whenever a state is changed from the
+ * frontend and whenever a state is actually changed and should be pushed to the frontend
+ */
+class ClimateDevice : public Nameable, public PollingComponent {
+ public:
+  /// Construct the climatedevice with name.
+  explicit ClimateDevice(const std::string &name, uint32_t update_interval = 60000);
+
+  /// Register a callback that will be called each time the mode changes is requested.
+  void add_on_state_callback(std::function<void(ClimateDeviceState)> &&callback);
+  /// Register a callback that will be called each time the mode changes.
+  void add_on_publish_state_callback(std::function<void()> &&callback);
+  /// Register a callback that will be called each time the current temperature changes
+  /// (parameter is target_temperature - current_temperature).
+  void add_on_current_temperature_callback(std::function<void(float)> &&callback);
+
+  /// Get the traits of this climate device (i.e. what features it supports).
+  const ClimateDeviceTraits &get_traits() const;
+  /// Set the traits of this climate device (i.e. what features it supports).
+  void set_traits(const ClimateDeviceTraits &traits);
+
+  template<typename... Ts> ControlAction<Ts...> *make_control_action();
+  template<typename... Ts> PublishAction<Ts...> *make_publish_action();
+  template<typename... Ts> CurrentTemperatureAction<Ts...> *make_current_temperature_action();
```

Comment:
Three actions:
- control: called if e.g. a state change is requested externally (automation or via MQTT). The requested state may needs to be processed.
- publish: To write and publish a new state, e.g. template component and internal state calculation
- current_temperature: To be called every time the current temperature is changed/polled. For PID controller a high sample rate may be required.